### PR TITLE
Continue dumping sections even when one is unknown

### PR DESCRIPTION
**Pull request by ereslibre**
This is relevant so that when we are dumping a module that contains
unknown sections that the current wabt tooling does not understand, we
still dump all the sections it understands, skipping (and printing an
error) about the unknown sections.

Before this change, we would print partial information, stopping at
the unknown section, without printing the rest of the sections that
wabt could potentially be able to understand.

Motivation: this happened when inspecting with `wasm-objdump` a module produced by [wasmlink](https://github.com/bytecodealliance/wit-bindgen/tree/main/crates/wasmlink).

Output when having produced a module with an unknown section with current version:

```
$ wasm-objdump -x linked.wasm

linked.wasm:	file format wasm 0x1
000020b: error: invalid section code: 14

Section Details:

Type[7]:
 - type[0] (i32, i32, i32, i32) -> i32
 - type[1] (i32, i32) -> i32
 - type[2] (i32) -> nil
 - type[3] (i32, i64, i32, i32) -> i32
 - type[4] (i32, i32, i32, i32, i32, i64, i64, i32, i32) -> i32
 - type[5] (i32, i32, i32) -> i32
 - type[6] (i32) -> i32
Import[12]:
 - func[0] sig=0 <wasi_snapshot_preview1.fd_write> <- wasi_snapshot_preview1.fd_write
 - func[1] sig=1 <wasi_snapshot_preview1.environ_get> <- wasi_snapshot_preview1.environ_get
 - func[2] sig=1 <wasi_snapshot_preview1.environ_sizes_get> <- wasi_snapshot_preview1.environ_sizes_get
 - func[3] sig=2 <wasi_snapshot_preview1.proc_exit> <- wasi_snapshot_preview1.proc_exit
 - func[4] sig=1 <wasi_snapshot_preview1.fd_filestat_get> <- wasi_snapshot_preview1.fd_filestat_get
 - func[5] sig=0 <wasi_snapshot_preview1.fd_read> <- wasi_snapshot_preview1.fd_read
 - func[6] sig=3 <wasi_snapshot_preview1.fd_seek> <- wasi_snapshot_preview1.fd_seek
 - func[7] sig=4 <wasi_snapshot_preview1.path_open> <- wasi_snapshot_preview1.path_open
 - func[8] sig=5 <wasi_snapshot_preview1.path_unlink_file> <- wasi_snapshot_preview1.path_unlink_file
 - func[9] sig=6 <wasi_snapshot_preview1.fd_close> <- wasi_snapshot_preview1.fd_close
 - func[10] sig=1 <wasi_snapshot_preview1.fd_prestat_get> <- wasi_snapshot_preview1.fd_prestat_get
 - func[11] sig=5 <wasi_snapshot_preview1.fd_prestat_dir_name> <- wasi_snapshot_preview1.fd_prestat_dir_name
```

As you can see, it stops as soon as it finds an unknown section. With the proposed patch, the output looks like:

```
$ wasm-objdump -x linked.wasm

linked.wasm:	file format wasm 0x1
000020b: error: invalid section code: 14
0377572: error: invalid section code: 15
03775c2: error: invalid section code: 16

Section Details:

Type[7]:
 - type[0] (i32, i32, i32, i32) -> i32
 - type[1] (i32, i32) -> i32
 - type[2] (i32) -> nil
 - type[3] (i32, i64, i32, i32) -> i32
 - type[4] (i32, i32, i32, i32, i32, i64, i64, i32, i32) -> i32
 - type[5] (i32, i32, i32) -> i32
 - type[6] (i32) -> i32
Import[12]:
 - func[0] sig=0 <wasi_snapshot_preview1.fd_write> <- wasi_snapshot_preview1.fd_write
 - func[1] sig=1 <wasi_snapshot_preview1.environ_get> <- wasi_snapshot_preview1.environ_get
 - func[2] sig=1 <wasi_snapshot_preview1.environ_sizes_get> <- wasi_snapshot_preview1.environ_sizes_get
 - func[3] sig=2 <wasi_snapshot_preview1.proc_exit> <- wasi_snapshot_preview1.proc_exit
 - func[4] sig=1 <wasi_snapshot_preview1.fd_filestat_get> <- wasi_snapshot_preview1.fd_filestat_get
 - func[5] sig=0 <wasi_snapshot_preview1.fd_read> <- wasi_snapshot_preview1.fd_read
 - func[6] sig=3 <wasi_snapshot_preview1.fd_seek> <- wasi_snapshot_preview1.fd_seek
 - func[7] sig=4 <wasi_snapshot_preview1.path_open> <- wasi_snapshot_preview1.path_open
 - func[8] sig=5 <wasi_snapshot_preview1.path_unlink_file> <- wasi_snapshot_preview1.path_unlink_file
 - func[9] sig=6 <wasi_snapshot_preview1.fd_close> <- wasi_snapshot_preview1.fd_close
 - func[10] sig=1 <wasi_snapshot_preview1.fd_prestat_get> <- wasi_snapshot_preview1.fd_prestat_get
 - func[11] sig=5 <wasi_snapshot_preview1.fd_prestat_dir_name> <- wasi_snapshot_preview1.fd_prestat_dir_name
Export[5]:
 - memory[0] -> "memory"
 - func[15] <_start> -> "_start"
 - func[16] <main> -> "main"
 - func[17] <canonical_abi_realloc> -> "canonical_abi_realloc"
 - func[18] <canonical_abi_free> -> "canonical_abi_free"
Elem[1]:
 - segment[0] flags=6 table=0 count=3 - init i32=0
  - elem[0] = func[12]
  - elem[1] = func[13]
  - elem[2] = func[14]
```

Thus, printing as much information as we can understand from the module. In terms of the C++ library the strict behavior and compatibility does not change, given `ReadBinaryOptions` was already exposed. We only make the case of `ReadBinaryOptions::stop_on_first_error` less strict when it's false.

### CODE REVIEW

File: src/binary-reader.cc
Comment (binji) [inline src/binary-reader.cc:70]: IMO we could keep this a little simpler by relying on brace-initializer syntax instead, see https://godbolt.org/z/je9sT5Eo5

**Review by binji — APPROVED** [review]
lgtm, this seems like a good fix. Thanks!

File: src/binary-reader.cc
Comment (ereslibre) [reply src/binary-reader.cc]: This has been addressed. I have ran clang-format to double check that the formatting is fine. I have not found any other linters `wabt` seems to use, so I'll have to wait for CI approval to confirm if everything is fine.

**Pull request merged**
